File: jython/buffer.py
import time 
import packet 
import threading
import globals
import defense
# import queue
import math
# import logging

# lock = threading.Lock()

# Buffer = queue.Queue()

def enqueuePacket(pkt):

  
   globals.LOCK_INGRESS_CAP[pkt.ingress].acquire()
   if ((globals.INGRESS_CAP[pkt.ingress].availableBuffSpace - pkt.packet_len) > 0):

      globals.LOCK_RECEIVE_COUNTER[pkt.ingress].acquire()
      globals.RECEIVE_COUNTER[pkt.ingress] +=1
      globals.BUFFER[pkt.ingress].put(pkt)
      globals.INGRESS_CAP[pkt.ingress].availableBuffSpace -= pkt.packet_len
      # globals.DEBUG_LOGGER.debug("Function: enqueuePacket - Packet Added to Queue, Available Buffer space at %(pkt.ingress) = %(globals.INGRESS_CAP[pkt.ingress].availableBuffSpace)" )
      globals.LOCK_RECEIVE_COUNTER[pkt.ingress].release()

   else:
      dropPacket(pkt)
   globals.LOCK_INGRESS_CAP[pkt.ingress].release()


def processPacket():
   global Buffer
   for i in range(0,globals.INGRESS_LOC):
      # globals.LOCK_RECEIVE_COUNTER[i].acquire()
      
      if(globals.RECEIVE_COUNTER[i] > 0):
         
         globals.LOCK_INGRESS_CAP[i].acquire()
         pktsToDequeue = globals.INGRESS_CAP[i].numOfDequeuePkts
         if(pktsToDequeue*globals.PKT_LEN > (globals.INGRESS_CAP[i].cap - globals.INGRESS_CAP[i].availableBuffSpace)):
            pktsToDequeue = math.floor((globals.INGRESS_CAP[i].cap - globals.INGRESS_CAP[i].availableBuffSpace)*1.0/globals.PKT_LEN)
         globals.INGRESS_CAP[i].availableBuffSpace += (globals.PKT_LEN*pktsToDequeue)
         
         # globals.DEBUG_LOGGER.debug(f"Function: processPacket - {pktsToDequeue} packets processed, Available Buffer space at {i} = {globals.INGRESS_CAP[i].availableBuffSpace}")
         globals.LOCK_INGRESS_CAP[i].release()

         for j in range(0,int(pktsToDequeue)):
            pkt = globals.BUFFER[i].get()
            defense.diagnose(pkt)

      

         globals.DEBUG_LOGGER.debug("Function: processPacket - after diagnose %s",str(globals.CURR_TRAFFIC_STATS[i]))
      
      # globals.LOCK_RECEIVE_COUNTER[i].release()
      

         
   
   
      # globals.processCounter[pkt.dst] += INGRESS_CAP[pkt.dst].numOfDequeuePkts


def dropPacket(pkt):
   if (pkt.attack_flag == 0):
      
      globals.LOCK_legitimateDropCounter[pkt.ingress].acquire()
      globals.legitimateDropCounter[pkt.ingress]+=1
      # globals.DEBUG_LOGGER.debug(f"Function: dropPacket - Legitimate packet dropped, legitimateDropCounter = {globals.legitimateDropCounter[pkt.ingress]}")
      globals.LOCK_legitimateDropCounter[pkt.ingress].release()

   else:
      globals.attackDropCounter[pkt.ingress] +=1
      # globals.DEBUG_LOGGER.debug(f"Function: dropPacket - Attack packet dropped, attackDropCounter = {globals.attackDropCounter[pkt.ingress]}")

File: go/src/simulator/buffer.go
package main

import (
	"math"
	"time"
	"fmt"
	
)

func enqueuePacket(pkt packet) {

	LOCK_INGRESS_CAP[pkt.ingress].Lock()
	if (INGRESS_CAP[pkt.ingress].availableBuffSpace - pkt.packet_len) > 0 {

		// LOCK_RECEIVE_COUNTER[pkt.ingress].Lock()
		// RECEIVE_COUNTER[pkt.ingress] +=1
		// LOCK_RECEIVE_COUNTER[pkt.ingress].Unlock()

		INGRESS_CAP[pkt.ingress].availableBuffSpace -= pkt.packet_len
		// _DEBUG.Printf("Function: enqueuePacket - Packet Added to Queue, Available Buffer space at %d = %f", pkt.ingress, INGRESS_CAP[pkt.ingress].availableBuffSpace)

		LOCK_INGRESS_CAP[pkt.ingress].Unlock()
		diagnose(pkt)
		// BUFFER[pkt.ingress].Add(pkt)
		// BUFFER[pkt.ingress] <- pkt   // Send v to channel ch

	} else {
		LOCK_INGRESS_CAP[pkt.ingress].Unlock()
		dropPacket(pkt)
	}

}

func dropPacket(pkt packet) {

	if pkt.attack_flag == false {

		LOCK_legitimateDropCounter[pkt.ingress].Lock()
		legitimateDropCounter[pkt.ingress] += 1
		// _DEBUG.Printf("Function: dropPacket - Legitimate packet dropped, legitimateDropCounter = %d",legitimateDropCounter[pkt.ingress])
		LOCK_legitimateDropCounter[pkt.ingress].Unlock()

		//  	} else {
		// 	attackDropCounter[pkt.ingress] +=1
		//      // # _DEBUG.Printf(f"Function: dropPacket - Attack packet dropped, attackDropCounter = {attackDropCounter[pkt.ingress]}")
	}
}

// func dequeuePackets(pktsToDequeue int, ingress int) {
// 	for j := 0 ; j < pktsToDequeue ; j++ {
// 				// pkt := <-BUFFER[i]
// 		pkt,ok  := (BUFFER[ingress].Next()).(packet)
// 		if(ok) {
// 			// enqueuePacket(pkt)
// 			diagnose(pkt)
// 		}
// 	}
// }
// func processing() {
// 	for {
// 		processPacket()
// 		// time.Sleep(10 * time.Microsecond)
// 	}
// }

func processPacket() {

    if BACKLOG==256{
        fmt.Println("Backlog Full")  
    }
    
    s:=time.Now()
    s.Add(time.Second*5)
    backlog[BACKLOG]=s.String()
    BACKLOG+=1
    
    st:=time.Now()
    for i:=0; i< BACKLOG; i++{
    
        times=st.String()        
        if (times==backlog[BACKLOG]){
        
            for j:=i; j<BACKLOG-1; j++{
                backlog[j]=backlog[j+1]
            }
            
            BACKLOG-=1
        }
   
     }
    

	for i := 0; i < CONFIGURATION.INGRESS_LOC; i++ {
		// # LOCK_RECEIVE_COUNTER[i].Lock()

		// if(RECEIVE_COUNTER[i] > 0) {

		LOCK_INGRESS_CAP[i].Lock()
		pktsToDequeue := int(math.Ceil((INGRESS_CAP[i].vmQueue - INGRESS_CAP[i].availableBuffSpace) / PKT_LEN))
		if pktsToDequeue > INGRESS_CAP[i].numOfDequeuePkts {
			pktsToDequeue = INGRESS_CAP[i].numOfDequeuePkts
		}
		// if(float64(pktsToDequeue)*PKT_LEN > (INGRESS_CAP[i].cap - INGRESS_CAP[i].availableBuffSpace)) {

		// }
		// _DEBUG.Printf("Function: processPacket - %d packets to be processed, Available Buffer space at %d = %f", pktsToDequeue, i, INGRESS_CAP[i].availableBuffSpace)

		INGRESS_CAP[i].availableBuffSpace += (PKT_LEN * float64(pktsToDequeue))

		// _DEBUG.Printf("Function: processPacket - %d packets processed, Available Buffer space at %d = %f", pktsToDequeue, i, INGRESS_CAP[i].availableBuffSpace)
		LOCK_INGRESS_CAP[i].Unlock()
		// go dequeuePackets(pktsToDequeue,i)

		// _DEBUG.Printf("Function: processPacket - after diagnose pkts to dequeue = %d", pktsToDequeue)

		// }

		// # LOCK_RECEIVE_COUNTER[i].Unlock()

	}
}
